core/src/main/java/io/grpc/internal/ClientTransportFactory.java

```diff
@@ -53,6 +55,15 @@ ConnectionClientTransport newClientTransport(
    */
   ScheduledExecutorService getScheduledExecutorService();
 
+  /**
+   * Swaps to a new ChannelCredentials with all other settings unchanged. Returns null if the
+   * ChannelCredentials is not supported by the current ClientTransportFactory settings, indicating
+   * a fallback to using the default ManagedChannelRegistry.
+   */
+  @CheckReturnValue
+  @Nullable
+  ClientTransportFactory withNewChannelCredential(ChannelCredentials channelCreds);
```

Comment:
I'd really rather one entity (the transport) fully-process the credentials by itself. As this stands, it is asking for trouble where the ManagedChannelImpl implementation doesn't agree with the `withNewChannelCredential()` implementation. For example, if you look at NettyChannelBuilder, the `ProtocolNegotiators.from()`'s result includes CallCredentials, but those are ignored (they should be null, but there's no check).

Could we please just return a result object that contains both the transport factory and the CallCredentials?

Reply:
Yes. Now I composite TransportFactory and CallCredentials as return type. It looks quite artificial but a lot convenient to propagate the data.